### PR TITLE
Add handling for ROSE_PYTHONPATH

### PR DESCRIPTION
**Pull request by wxtim**
Exactly the same as https://github.com/cylc/cylc-flow/pull/5727

### CODE REVIEW

**Review by oliver-sanders — COMMENTED** [review]
LGTM, we'll need to work out how to advertise thus one.

**Comment by wxtim** [comment]
> LGTM, we'll need to work out how to advertise thus one.

Do you think I should try adding a doc change to this PR?

File: metomi/rose/rose.py [hunk 1]
```diff
@@ -14,11 +14,39 @@
 #
 # You should have received a copy of the GNU General Public License
 # along with Rose. If not, see <http://www.gnu.org/licenses/>.
+
+import os
+import sys
+
+
+def pythonpath_manip():
+    """Stop PYTHONPATH contaminating the Cylc Environment
```
Comment (oliver-sanders) [inline metomi/rose/rose.py:23]:
```suggestion
    """Stop PYTHONPATH contaminating the Rose Environment
```

File: tox.ini [hunk 2]
```diff
@@ -13,3 +13,5 @@ ignore=
     E731,
     # no longer best practice:
     W503
+    ; module level import not at top of file
+    E402,
```
Comment (oliver-sanders) [inline tox.ini:17]: Might be able to do this with inline comments rather than globally.

**Review by oliver-sanders — APPROVED** [review]

File: tox.ini (hunk 2 above)
Comment (wxtim) [reply tox.ini:17]: Was coping the Cylc `tox.ini`. We don't generally do this, but when we do we usually have our reasons. It's not a fault I think we'll miss.

File: CHANGES.md [hunk 3]
```diff
@@ -13,6 +13,13 @@ ones in. -->
 
 ## 2.1.0 (<span actions:bind='release-date'>Released 2023-07-21</span>)
 
+### Breaking Changes
+
+[2736](https://github.com/metomi/rose/pull/2736)
```
Comment (MetRonnie) [inline CHANGES.md:18]: Under the wrong heading

File: metomi/rose/rose.py [hunk 4]
```diff
@@ -14,11 +14,39 @@
...
+    * Add items from ROSE_PYTHONPATH to sys.path.
+
+    See Also:
+        https://github.com/cylc/cylc-flow/issues/5124
+    """
+    if 'ROSE_PYTHONPATH' in os.environ:
+        for item in os.environ['ROSE_PYTHONPATH'].split(os.pathsep):
+            print(f'extracted {item} from ROSE_PYTHONPATH')
+            abspath = os.path.abspath(item)
+            sys.path.insert(0, abspath)
```
Comment (MetRonnie) [inline metomi/rose/rose.py:36]: Aware this is the same logic added in Cylc, but I've just noticed this reverses the order set by the user?

File: metomi/rose/tests/test_rose.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    monkeypatch.setenv('PYTHONPATH', '/remove-from-sys.path')
+    monkeypatch.setattr('sys.path', ['/leave-alone', '/remove-from-sys.path'])
+    pythonpath_manip()
+    # ... we don't change PYTHONPATH
+    assert os.environ['PYTHONPATH'] == '/remove-from-sys.path'
+    # ... but we do remove PYTHONPATH items from sys.path, and don't remove
+    # items there not in PYTHONPATH
+    assert sys.path == ['/leave-alone']
+
+    # If CYLC_PYTHONPATH is set we retrieve its contents and
```
Comment (MetRonnie) [inline metomi/rose/tests/test_rose.py:40]:
```suggestion
    # If ROSE_PYTHONPATH is set we retrieve its contents and
```

**Review by MetRonnie — COMMENTED** [review]
Any idea why the tests are failing on 3.7?

File: metomi/rose/rose.py (hunk 4 above)
Comment (oliver-sanders) [reply metomi/rose/rose.py:36]: Ach, dammit.

Comment (MetRonnie) [reply metomi/rose/rose.py:36]:
```suggestion
        rose_pythonpath_vars = [
            os.path.abspath(item)
            for item in os.environ['ROSE_PYTHONPATH'].split(os.pathsep)
        ]
        print(f"Extracted {rose_pythonpath_vars} from ROSE_PYTHONPATH")
        sys.path.insert(0, *rose_pythonpath_vars)

```

File: metomi/rose/tests/test_rose.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    # add them to the sys.path:
+    monkeypatch.setenv('ROSE_PYTHONPATH', '/add-to-sys.path')
+    pythonpath_manip()
+    assert sys.path == ['/add-to-sys.path', '/leave-alone']
```
Comment (MetRonnie) [inline metomi/rose/tests/test_rose.py:44]:
```suggestion
    monkeypatch.setenv('PYTHONPATH', '/remove1:/remove2')
    monkeypatch.setattr('sys.path', ['/leave-alone', '/remove1', '/remove2'])
    pythonpath_manip()
    # ... we don't change PYTHONPATH
    assert os.environ['PYTHONPATH'] == '/remove1:/remove2'
    # ... but we do remove PYTHONPATH items from sys.path, and don't remove
    # items there not in PYTHONPATH
    assert sys.path == ['/leave-alone']

    # If ROSE_PYTHONPATH is set we retrieve its contents and
    # add them to the sys.path:
    monkeypatch.setenv('ROSE_PYTHONPATH', '/add1:/add2')
    pythonpath_manip()
    assert sys.path == ['/add1', '/add2', '/leave-alone']
```

File: CHANGES.md (hunk 3 above)
Comment (wxtim) [reply CHANGES.md:18]: If you are relying on `PYTHONPATH` to add functions for rose ana this change will break your workflow. On one side of the change you must set `PYTHONPATH`, on the other `ROSE_PYTHONPATH`.

File: CHANGES.md [hunk 7]
```diff
@@ -11,6 +11,15 @@ ones in. -->
 
 --------------------------------------------------------------------------------
 
+## 2.2.0 (<span actions:bind='release-date'>Upcoming</span>)
+
+### Breaking Changes
+
+[2736](https://github.com/metomi/rose/pull/2736)
```
Comment (MetRonnie) [inline CHANGES.md:18]:
```suggestion
[#2736](https://github.com/metomi/rose/pull/2736) -
```

**Comment by MetRonnie** [comment]
> Any idea why the tests are failing on 3.7?

`/rose-task-run/44-bunch-incremental.t` now failing on the others too 😢

**Review by MetRonnie — APPROVED** [review]
Can you follow-up on cylc-flow with the ordering fix

File: metomi/rose/rose.py [hunk 8]
```diff
@@ -14,11 +14,43 @@
...
+
+    See Also:
+        https://github.com/cylc/cylc-flow/issues/5124
+    """
+    if 'ROSE_PYTHONPATH' in os.environ:
+        paths = [
+            os.path.abspath(item)
+            for item in os.environ['ROSE_PYTHONPATH'].split(os.pathsep)
+        ]
+        print(f"Extracted {paths} from ROSE_PYTHONPATH")
```
Comment (wxtim) [inline metomi/rose/rose.py:37]:
```suggestion
```

**Review by MetRonnie — APPROVED** [review]

**Comment by MetRonnie** [comment]
`/rose-task-run/44-bunch-incremental.t` also failing flakily on master

**Pull request merged**